File: common/local_tools_compat.py
"""VMD / VESTA 文件类型支持矩阵 (前后端共用单例)。

事实来源:
  - VMD 文档支持: lammpstrj, dcd, xtc, trr, xyz, pdb, gro 等分子/晶体轨迹
  - VESTA 文档支持: vesta, cif, xsf, xyz, pdb, cssr, struct, xod, xtl, fdf
  - VESTA **不支持** LAMMPS .data 格式 (会弹 "Invalid data! This file type is not supported.")

判断函数供前端按扩展名筛按钮、后端 spawn 前再次校验 (.data 自动转 .xyz 临时文件)。
"""
from __future__ import annotations

VMD_EXTS: frozenset[str] = frozenset({
    ".lammpstrj", ".dump", ".dcd", ".xtc", ".trr", ".xyz", ".pdb", ".gro",
})

VESTA_EXTS: frozenset[str] = frozenset({
    ".vesta", ".cif", ".xsf", ".xyz", ".pdb", ".cssr", ".struct", ".xod", ".xtl", ".fdf",
})


def supports(tool: str, ext: str) -> bool:
    """扩展名是否被指定工具原生支持 (不经过转换)。"""
    ext = ext.lower()
    if not ext.startswith("."):
        ext = "." + ext
    if tool == "vmd":
        return ext in VMD_EXTS
    if tool == "vesta":
        return ext in VESTA_EXTS
    return False

File: workbench/frontend/src/utils/localToolsCompat.ts
/** 前端 VMD/VESTA 扩展名支持矩阵 (与 common/local_tools_compat.py 同源, 手动同步)。
 *  后端是真相源, 这里只供前端按扩展名筛按钮; 后端 spawn 前会再次校验。 */

const VMD_EXTS = new Set([
  ".lammpstrj", ".dump", ".dcd", ".xtc", ".trr", ".xyz", ".pdb", ".gro",
]);
const VESTA_EXTS = new Set([
  ".vesta", ".cif", ".xsf", ".xyz", ".pdb", ".cssr", ".struct", ".xod", ".xtl", ".fdf",
]);

export function supports(tool: "vmd" | "vesta", ext: string): boolean {
  const e = ext.toLowerCase();
  if (!e.startsWith(".")) return false;
  if (tool === "vmd") return VMD_EXTS.has(e);
  if (tool === "vesta") return VESTA_EXTS.has(e);
  return false;
}
